Guard against `string.Concat(object)` returning null

canton7 opened the pull request:
This PR is intended as a starting point for any discussion on #34085.

Note that this is a **BREAKING CHANGE** (although it's a correction to align with the spec, it's very niche, and a similar case was inadvertently fixed by Roslyn and I don't believe anyone noticed).

I believe it stands on its own as a working solution (for C# - it doesn't (yet) cover VB), but anticipate that it may need to change, or be rejected entirely because of its breaking nature.

Alternatives to this approach:

1. Change `string.Concat(object)` to never return null (also a breaking change)
2. Use `o?.ToString() ?? ""` instead of `string.Concat(o) ?? ""`
3. Change the spec, or accept and live with the deviation

------------------------------

If you define an object with a ToString method which returns null:

    public class C
    {
        public override string ToString() => null;
    }

And then concatenate that with `null` or "":

    C c = new C();
    c + ""

then the compiler emits:

    string.Concat(c)

However, `string.Concat(object)` will return null if the object's
ToString method returns null. This contradicts section 7.8.4 of the
Specification, which says:

    If ToString returns null, an empty string is substituted.
    ...
    The string concatenation operator never returns a null value

On the other hand, if you write:

    c + "" + ""

then the compiler emits:

    string.Concat(c) ?? ""

This is because it reduces `c + ""` to `string.Concat(c)` as before, but
when it considers the second `+ ""`, it sees `<string> + ""` and simplifies
that to `<string> ?? ""`. This is unnecessary in this case, but
inadvertently fixes this case.

This PR always emits `string.Concat(c) ?? ""`. It enhances the logic
which extracts the arguments to a call to `string.Concat` so it can unwrap
the null coalescing operator, and it enhances the `<string> + ""` logic
so that it only adds `?? ""` if `<string>` is not a call to `string.Concat`
(or `string.Concat(object) ?? ""`), since we know that this can never
result in null.

This introduces one more allocation: a single-element ImmutableArray
in some cases.

It seems that compiled expressions were never affected by this. Their
behaviour now matches that of compiled C#.

This does not change any of the VB code. I haven't looked into what the
VB spec says here, and I want any discussions on this PR to be resolved
before I invest that effort.

Fixes #34085

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/roslyn?pullRequest=34208) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/roslyn?pullRequest=34208) before we can accept your contribution.<br/><br/>:x: canton7 [sign now](https://cla.dotnetfoundation.org/dotnet/roslyn?pullRequest=34208)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/roslyn?pullRequest=34208) it.</sub>

canton7 commented:
The test failure is because the test case tests what happens if `string.Concat(object)` is missing, and so [this line](https://github.com/dotnet/roslyn/pull/34208/files#diff-4ba33044f27f4544fb3d1ddf4903e749R295) gets an `ErrorMethodSymbol`, which it then passes to `_factoryCoalesce`, which hits a `Debug.Assert`.

What's the correct way to handle this? Should I do an explicit check to see if `method` is an `ErrorMethodSymbol` before trying to pass it to `BoundCall.Synthesized` and subsequently `_factory.Coalesce`?

gafter commented:
Continuing discussion of possible compat concerns at https://github.com/dotnet/roslyn/issues/34085

RikkiGibson commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenStringConcat.cs:
```diff
@@ -358,18 +358,84 @@ static void Main()
             comp.VerifyIL("Test.Main", @"
 {
-  // Code size       35 (0x23)
+  // Code size       44 (0x2c)
```
how does this compare to the IL produced for the pre-Roslyn compiler?

canton7 replied:
In this specific case, pre-Roslyn generates the same as Roslyn did before this PR: a straight call to `string.Concat(object)` with no subsequent null-check for an object, and `S ?? ""` for a string.

Pre-Roslyn generates the same for `O + null + null` and `S + null + null` (whereas Roslyn generated `string.Concat(O) ?? ""` for the first case).

gafter commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs:
```diff
@@ -257,14 +281,22 @@ private BoundExpression RewriteStringConcatenationOneExpr(SyntaxNode syntax, Bou
             {
-                // loweredOperand ?? ""
-                return _factory.Coalesce(loweredOperand, _factory.Literal(""));
+                // It it's a call to string.Concat, we already know the result is not null. Otherwise return loweredOperand ?? ""
```
>It it's a call to string.Concat, we already know the result is not null [](start = 19, length = 71)

This appears to be incorrect.  You point out in the OP a circumstance in which `string.Concat` can return null.

canton7 replied:
Right. The compiler will never emit `string.Concat(object)` on its own (it always emits `string.Concat(object) ?? ""`), but of course the user can compile the code `string.Concat(c) + null`, and that will get compiled to `string.Concat(c)`. 

I added this optimization to avoid `o + null + null` from being compiled to `string.Concat(o) ?? "" ?? ""` (which just looks like a bug to anyone inspecting the compiler output), and `o + o + ""` from being compiled to `string.Concat(o, o) ?? ""` (which is an optimization that the pre-Roslyn compiler does, but it seems it was lost in Roslyn).

I think there's a way to fix this cleanly. I'll put together a commit later, then please let me know whether you want to keep it with the fix, or remove it altogether.

gafter commented:
We are OK with the compat implications of this.

Suchiman commented:
From what i can tell, only `String.Concat(object)` may return null, so the optimization is still valid for the other overloads.

gafter commented on src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs:
```diff
@@ -717,7 +717,7 @@ public BoundExpression ComplexConditionalReceiver(BoundExpression valueTypeRecei
         public BoundExpression Coalesce(BoundExpression left, BoundExpression right)
         {
-            Debug.Assert(left.Type.Equals(right.Type, TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds | TypeCompareKind.IgnoreNullableModifiersForReferenceTypes));
+            Debug.Assert(left.Type.Equals(right.Type, TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds | TypeCompareKind.IgnoreNullableModifiersForReferenceTypes) || left.Type.IsErrorType());
```
>eft.Type.IsErrorType() [](start = 183, length = 22)

The need for this confuses me.  If a program has errors, we don't lower it or perform code generation.  So why would this be needed?

canton7 replied:
This assertion was triggered by the unit test which tests the case where `string.Concat(object)` is missing. I'm not at all convinced that this is the right approach: please let me know if I should do something else.

gafter replied:
I understand: the error bound node is triggered by an error arising during lowering.  So this is correct.

gafter commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs:
```diff
@@ -168,16 +152,73 @@ private void FlattenConcatArg(BoundExpression lowered, ArrayBuilder<BoundExpress
-            // fallback - if nothing above worked, leave arg as-is
-            flattened.Add(lowered);
-            return;
+            arguments = ImmutableArray<BoundExpression>.Empty;
```
>ImmutableArray<BoundExpression>.Empty [](start = 24, length = 37)

default

canton7 commented:
I'm afraid I seem to have broken some of the IntelliSense tests, but I can't reproduce them locally and I'm not sure how to get details of the failure (stack trace, etc) out of roslyn-CI. Any guidance would be appreciated.

RikkiGibson commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs:
```diff
@@ -168,16 +152,73 @@ private void FlattenConcatArg(BoundExpression lowered, ArrayBuilder<BoundExpress
+                                return true;
+                            }
+
+                            arguments = ImmutableArray<BoundExpression>.Empty.Add(boundCoalesce.LeftOperand);
```
nit: consider `ImmutableArray.Create(boundCoalesce.LeftOperand)` here

canton7 replied:
Aha!

RikkiGibson commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs:
```diff
@@ -168,16 +152,73 @@ private void FlattenConcatArg(BoundExpression lowered, ArrayBuilder<BoundExpress
+
+        private bool TryExtractStringConcatArgsFromBoundCall(BoundCall boundCall, out ImmutableArray<BoundExpression> arguments, out bool callCanReturnNull)
+        {
+            // callCanReturnNull is true if this method returns true, and boundCall is a call to a string.Concat overload which can return null
```
Consider putting this comment which seems to describe the "contract" of this method above the signature. Same with `TryExtractStringConcatArgs`.

gafter commented:
@canton7 This has been approved to integrate.  Please tell us when you're ready for us to merge it (i.e. let us know if you want to make any further changes based on the most recent review comments).

canton7 commented:
Thanks! I'll make some changes tomorrow in response to the good points that @RikkiGibson raised, and let you know when I'm done.

canton7 commented:
I've implemented @RikkiGibson's comments, rebased onto master, and tidied up the commits. After the rebase the whole test suite now passes, with the exception of some of the Windows_VisualStudio_Integration_Tests tests (which seem to be failing for other PRs as well).

Once the test suite passes, I think we're good to go!

gafter commented:
@canton7 In the future please do not rebase.  It makes it impossible for us to determine what was changed since we last reviewed.  We usually rebase and squash all commits when we merge.

gafter commented:
@canton7 Thank you so much for your contribution!

canton7 commented:
Apologies, I didn't see anything about it on [Contributing Code](https://github.com/dotnet/roslyn/wiki/Contributing-Code).

Thanks everyone for your time getting this through! I appreciate the support and effort.
